yarn-project/end-to-end/src/uniswap_trade_on_l1_from_l2.test.ts

```diff
@@ -407,5 +427,316 @@ describe('uniswap_trade_on_l1_from_l2', () => {
+
+  it("uniswap contract can't unshield if I don't give the right approval", async () => {
+    // swap should fail since no withdraw approval to uniswap:
+    const nonceForWETHUnshieldApproval = new Fr(2n);
+
+    const expectedMessageHash = await computeAuthWitMessageHash(
+      uniswapL2Contract.address,
+      wethCrossChainHarness.l2Token.methods
+        .unshield(ownerAddress, uniswapL2Contract.address, wethAmountToBridge, nonceForWETHUnshieldApproval)
+        .request(),
+    );
+
+    await expect(
+      uniswapL2Contract.methods
+        .swap_private(
+          wethCrossChainHarness.l2Token.address,
+          wethCrossChainHarness.l2Bridge.address,
+          wethAmountToBridge,
+          daiCrossChainHarness.l2Bridge.address,
+          nonceForWETHUnshieldApproval,
+          uniswapFeeTier,
+          minimumOutputAmount,
+          Fr.random(),
+          Fr.random(),
+          deadlineForDepositingSwappedDai,
+          ownerEthAddress,
+          ownerEthAddress,
+        )
+        .simulate(),
+    ).rejects.toThrowError(`Unknown auth witness for message hash 0x${expectedMessageHash.toString('hex')}`);
+  });
+
+  it("can't swap if user passes a token different to what the bridge tracks", async () => {
+    // 1. give user private funds on L2:
+    const [secretForRedeemingWeth, secretHashForRedeemingWeth] = await wethCrossChainHarness.generateClaimSecret();
+    await wethCrossChainHarness.mintTokensPrivateOnL2(wethAmountToBridge, secretHashForRedeemingWeth);
+    await wethCrossChainHarness.redeemShieldPrivatelyOnL2(wethAmountToBridge, secretForRedeemingWeth);
+    await wethCrossChainHarness.expectPrivateBalanceOnL2(ownerAddress, wethAmountToBridge);
+
+    // 2. owner gives uniswap approval to unshield funds:
+    logger('Approving uniswap to unshield funds to self on my behalf');
+    const nonceForWETHUnshieldApproval = new Fr(3n);
+    const unshieldToUniswapMessageHash = await computeAuthWitMessageHash(
+      uniswapL2Contract.address,
+      wethCrossChainHarness.l2Token.methods
+        .unshield(ownerAddress, uniswapL2Contract.address, wethAmountToBridge, nonceForWETHUnshieldApproval)
+        .request(),
+    );
+    await ownerWallet.createAuthWitness(Fr.fromBuffer(unshieldToUniswapMessageHash));
+
+    // 3. Swap but send the wrong token address
+    logger('Swap but send the wrong token address');
+    await expect(
+      uniswapL2Contract.methods
+        .swap_private(
+          wethCrossChainHarness.l2Token.address, // send weth token
+          daiCrossChainHarness.l2Bridge.address, // but dai bridge!
+          wethAmountToBridge,
+          daiCrossChainHarness.l2Bridge.address,
+          nonceForWETHUnshieldApproval,
+          uniswapFeeTier,
+          minimumOutputAmount,
+          Fr.random(),
+          Fr.random(),
+          deadlineForDepositingSwappedDai,
+          ownerEthAddress,
+          ownerEthAddress,
+        )
+        .simulate(),
+    ).rejects.toThrowError(
```

Comment:
The error message is a little messy here. I have been able to work around it on some other cases by having a `== true`. Not sure if it would work here to make the assertion
```rust
assert((token == TokenBridge::at(token_bridge).token(context)) == true, "input_asset address is not the same as seen in the bridge contract");
```

Reply:
ooh good idea!

Reply:
actually I am going to use AztecAddress.eq which somehow also creates a good error mesage